nighthawk/source/client/benchmark_http_client.cc

```diff
@@ -0,0 +1,192 @@
+#include "client/benchmark_http_client.h"
+
+#include "common/http/utility.h"
+#include "common/network/utility.h"
+
+#include "ares.h"
+
+#include "absl/strings/str_split.h"
+
+#include "common/api/api_impl.h"
+#include "common/common/compiler_requirements.h"
+#include "common/event/dispatcher_impl.h"
+#include "common/http/header_map_impl.h"
+#include "common/http/headers.h"
+#include "common/http/http1/conn_pool.h"
+#include "common/http/http2/conn_pool.h"
+#include "common/network/raw_buffer_socket.h"
+#include "common/network/utility.h"
+#include "common/runtime/runtime_impl.h"
+#include "common/thread_local/thread_local_impl.h"
+#include "common/upstream/cluster_manager_impl.h"
+
+#include "common/ssl.h"
+#include "common/stream_decoder.h"
+
+using namespace std::chrono_literals;
+
+namespace Nighthawk {
+namespace Client {
+
+BenchmarkHttpClient::BenchmarkHttpClient(Envoy::Event::Dispatcher& dispatcher,
+                                         Envoy::Stats::Store& store, Envoy::TimeSource& time_source,
+                                         const std::string& uri,
+                                         Envoy::Http::HeaderMapImplPtr&& request_headers,
+                                         bool use_h2)
+    : dispatcher_(dispatcher), store_(store), time_source_(time_source),
+      request_headers_(std::move(request_headers)), use_h2_(use_h2), is_https_(false), host_(""),
+      port_(0), path_("/"), dns_failure_(true), timeout_(5s), connection_limit_(1),
+      max_pending_requests_(1), pool_overflow_failures_(0), stream_reset_count_(0),
+      http_good_response_count_(0), http_bad_response_count_(0), requests_completed_(0),
+      requests_initiated_(0), allow_pending_for_test_(false) {
+
+  // parse incoming uri into fields that we need.
+  // TODO(oschaaf): refactor. also input validation, etc.
+  absl::string_view host, path;
+  Envoy::Http::Utility::extractHostPathFromUri(uri, host, path);
+  host_ = std::string(host);
+  path_ = std::string(path);
+
+  size_t colon_index = host_.find(':');
+  is_https_ = absl::StartsWith(uri, "https://");
+
+  if (colon_index == std::string::npos) {
+    port_ = is_https_ ? 443 : 80;
```

Comment:
Not sure if we need to second guess the user here; if they say HTTPS as the URI, let's do that, otherwise stick to HTTP.

Reply:
I was hoping that this would determine the default port we should use based on the protocol the user specified, in case he didn't explitly specify the port?